src/dnmd/deltas.c

```diff
@@ -1,125 +1,314 @@
         return false;
 
-    // Copy over base and delta
-    memcpy(mem->data, base->ptr, base->size);
-    memcpy(mem->data + base->size, delta->ptr + delta_offset, delta_size);
+    if (!append_heap(cxt, delta, mdtc_hus))
+        return false;
 
-    // Update base with new memory allocation
-    base->ptr = mem->data;
-    base->size = mem->size;
     return true;
 }
 
-static bool merge_string_heap(mdcxt_t* cxt, mdcxt_t* delta)
+typedef enum
 {
-    if (delta->strings_heap.size == 0)
-        return true;
+    dops_Default = 0,
+    dops_MethodCreate,
+    dops_FieldCreate,
+    dops_ParamCreate,
+    dops_PropertyCreate,
+    dops_EventCreate,
+} delta_ops_t;
 
-    mdstream_t* base = &cxt->strings_heap;
-    mdstream_t* str = &delta->strings_heap;
-    return merge_into_base(cxt, base, str, 0);
-}
+struct map_table_group_t
+{
+    mdcursor_t start;
+    uint32_t count;
+};
 
-static bool merge_guid_heap(mdcxt_t* cxt, mdcxt_t* delta)
+typedef struct _enc_token_map_t
 {
-    if (delta->guid_heap.size == 0)
+    struct map_table_group_t map_cur_by_table[MDTABLE_MAX_COUNT];
+} enc_token_map_t;
+
+static bool initialize_token_map(mdtable_t* map, enc_token_map_t* token_map)
+{
+    assert(map->table_id == mdtid_ENCMap);
+    for (uint32_t i = 0; i < MDTABLE_MAX_COUNT; ++i)
+    {
+        token_map->map_cur_by_table[i].count = (uint32_t)(-1);
+    }
+
+    // If we don't have any entries in the map table, then we don't have any remapped tokens.
+    // The initialization we've already done by this point is sufficient.
+    if (map->row_count == 0)
         return true;
 
-    mdstream_t* base = &cxt->guid_heap;
-    mdstream_t* guid = &delta->guid_heap;
+    // The EncMap table is grouped by token type and sorted by the order of the rows in the tables in the delta.
+    mdcursor_t map_cur = create_cursor(map, 1);
 
-    // The delta's GUID heap is never fully copied.
-    // Rather we merge only the newer portion of the heap.
-    return merge_into_base(cxt, base, guid, base->size);
-}
+    mdtable_id_t previous_table_id = mdtid_Unused;
+    for (uint32_t i = 0; i < map->row_count; (void)md_cursor_next(&map_cur), ++i)
+    {
+        mdToken tk;
+        if (1 != md_get_column_value_as_constant(map_cur, mdtENCMap_Token, 1, &tk))
+        {
+            return false;
+        }
+        
+        // Tokens in the EncMap table may have the high bit set to indicate that they aren't true token references.
+        // Deltas produced by CoreCLR, CLR, and ILASM will have this bit set. Roslyn does not utilize this bit.
+        // Strip this high bit if it is set. We don't need it.
+        mdtable_id_t table_id = ExtractTokenType(tk) & 0x7F;
 
-static bool merge_blob_heap(mdcxt_t* cxt, mdcxt_t* delta)
-{
-    if (delta->blob_heap.size == 0)
-        return true;
+        if (table_id < mdtid_First || table_id >= mdtid_End)
+            return false;
+
+        if (token_map->map_cur_by_table[table_id].count == (uint32_t)(-1))
```

Comment:
Is the `-1` a sentinel value that we need to encode somewhere?

Reply:
If so, I think a more appropriate definition would be in hex.

Reply:
I was following the pattern of `std::string::npos`, which uses `-1` cast to a `size_t`. I can change it to hex if desired.

Reply:
No that is fine. I think a const value somewhere for set/get makes sense.